// angulartics2/gst/bundles/angulartics2-gst.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('angulartics2')) :
    typeof define === 'function' && define.amd ? define('angulartics2/gst', ['exports', '@angular/core', 'angulartics2'], factory) :
    (factory((global.angulartics2 = global.angulartics2 || {}, global.angulartics2.gst = {}),global.ng.core,global.angulartics2));
}(this, (function (exports,core,angulartics2) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __values(o) {
        var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
        if (m)
            return m.call(o);
        return {
            next: function () {
                if (o && i >= o.length)
                    o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var GoogleGlobalSiteTagDefaults = (function () {
        function GoogleGlobalSiteTagDefaults() {
            var _this = this;
            this.trackingIds = [];
            if (typeof ga !== 'undefined' && ga) {
                // See: https://developers.google.com/analytics/devguides/collection/analyticsjs/ga-object-methods-reference
                ga.getAll().forEach(function (tracker) {
                    var /** @type {?} */ id = tracker.get('trackingId');
                    if (id !== undefined) {
                        _this.trackingIds.push(id);
                    }
                });
            }
        }
        return GoogleGlobalSiteTagDefaults;
    }());
    var Angulartics2GoogleGlobalSiteTag = (function () {
        function Angulartics2GoogleGlobalSiteTag(angulartics2$$1) {
            var _this = this;
            this.angulartics2 = angulartics2$$1;
            var /** @type {?} */ defaults = new GoogleGlobalSiteTagDefaults;
            // Set the default settings for this module
            this.angulartics2.settings.gst = __assign({}, defaults, this.angulartics2.settings.gst);
            this.angulartics2.pageTrack
                .pipe(this.angulartics2.filterDeveloperMode())
                .subscribe(function (x) { return _this.pageTrack(x.path); });
            this.angulartics2.eventTrack
                .pipe(this.angulartics2.filterDeveloperMode())
                .subscribe(function (x) { return _this.eventTrack(x.action, x.properties); });
            this.angulartics2.exceptionTrack
                .pipe(this.angulartics2.filterDeveloperMode())
                .subscribe(function (x) { return _this.exceptionTrack(x); });
        }
        /**
         * Manually track page view, see:
         *
         * https://developers.google.com/analytics/devguides/collection/gtagjs/single-page-applications#tracking_virtual_pageviews
         *
         * @param path relative url
         */
        /**
         * Manually track page view, see:
         *
         * https://developers.google.com/analytics/devguides/collection/gtagjs/single-page-applications#tracking_virtual_pageviews
         *
         * @param {?} path relative url
         * @return {?}
         */
        Angulartics2GoogleGlobalSiteTag.prototype.pageTrack = /**
         * Manually track page view, see:
         *
         * https://developers.google.com/analytics/devguides/collection/gtagjs/single-page-applications#tracking_virtual_pageviews
         *
         * @param {?} path relative url
         * @return {?}
         */
            function (path) {
                if (typeof gtag !== 'undefined' && gtag) {
                    try {
                        for (var _a = __values(this.angulartics2.settings.gst.trackingIds), _b = _a.next(); !_b.done; _b = _a.next()) {
                            var id = _b.value;
                            gtag('config', id, { 'page_path': path });
                        }
                    }
                    catch (e_1_1) {
                        e_1 = { error: e_1_1 };
                    }
                    finally {
                        try {
                            if (_b && !_b.done && (_c = _a.return))
                                _c.call(_a);
                        }
                        finally {
                            if (e_1)
                                throw e_1.error;
                        }
                    }
                }
                var e_1, _c;
            };
        /**
         * Send interactions to gtag, i.e. for event tracking in Google Analytics. See:
         *
         * https://developers.google.com/analytics/devguides/collection/gtagjs/events
         *
         * @param action associated with the event
         */
        /**
         * Send interactions to gtag, i.e. for event tracking in Google Analytics. See:
         *
         * https://developers.google.com/analytics/devguides/collection/gtagjs/events
         *
         * @param {?} action associated with the event
         * @param {?} properties
         * @return {?}
         */
        Angulartics2GoogleGlobalSiteTag.prototype.eventTrack = /**
         * Send interactions to gtag, i.e. for event tracking in Google Analytics. See:
         *
         * https://developers.google.com/analytics/devguides/collection/gtagjs/events
         *
         * @param {?} action associated with the event
         * @param {?} properties
         * @return {?}
         */
            function (action, properties) {
                // TODO: make interface
                //  @param {string} properties.category
                //  @param {string} [properties.label]
                //  @param {number} [properties.value]
                //  @param {boolean} [properties.noninteraction]
                // Set a default GST category
                properties = properties || {};
                if (typeof gtag !== 'undefined' && gtag) {
                    gtag('event', action, __assign({ event_category: properties.category || 'interaction', event_label: properties.label, value: properties.value, non_interaction: properties.noninteraction }, properties.gstCustom));
                }
            };
        /**
         * Exception Track Event in GST. See:
         *
         * https://developers.google.com/analytics/devguides/collection/gtagjs/exceptions
         *
         */
        /**
         * Exception Track Event in GST. See:
         *
         * https://developers.google.com/analytics/devguides/collection/gtagjs/exceptions
         *
         * @param {?} properties
         * @return {?}
         */
        Angulartics2GoogleGlobalSiteTag.prototype.exceptionTrack = /**
         * Exception Track Event in GST. See:
         *
         * https://developers.google.com/analytics/devguides/collection/gtagjs/exceptions
         *
         * @param {?} properties
         * @return {?}
         */
            function (properties) {
                // TODO: make interface
                //  @param {Object} properties
                //  @param {string} [properties.description]
                //  @param {boolean} [properties.fatal]
                if (properties.fatal === undefined) {
                    console.log('No "fatal" provided, sending with fatal=true');
                    properties.fatal = true;
                }
                properties.exDescription = properties.event ? properties.event.stack : properties.description;
                this.eventTrack('exception', {
                    gstCustom: __assign({ 'description': properties.exDescription, 'fatal': properties.fatal }, properties.gstCustom)
                });
            };
        Angulartics2GoogleGlobalSiteTag.decorators = [
            { type: core.Injectable }
        ];
        /** @nocollapse */
        Angulartics2GoogleGlobalSiteTag.ctorParameters = function () {
            return [
                { type: angulartics2.Angulartics2 }
            ];
        };
        return Angulartics2GoogleGlobalSiteTag;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    exports.GoogleGlobalSiteTagDefaults = GoogleGlobalSiteTagDefaults;
    exports.Angulartics2GoogleGlobalSiteTag = Angulartics2GoogleGlobalSiteTag;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
