// angulartics2/ga-enhanced-ecom/esm2015/ga-enhanced-ecom-options.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Product and Promotion Actions
 *
 * Actions specify how to interpret product and promotion data that you send to Google Analytics.
 */
/**
 * Impression Data
 *
 * Represents information about a product that has been viewed. It is
 * referred to as an impressionFieldObject and contains the following values.
 *
 * @link https://developers.google.com/analytics/devguides/collection/analyticsjs/enhanced-ecommerce#impression-data
 * @record
 */
export function GaEnhancedEcomImpressionFieldObject() { }
function GaEnhancedEcomImpressionFieldObject_tsickle_Closure_declarations() {
    /**
     * The product ID or SKU (e.g. P67890). *Either this field or name must be set.
     * @type {?}
     */
    GaEnhancedEcomImpressionFieldObject.prototype.id;
    /**
     * The name of the product (e.g. Android T-Shirt). *Either this field or id must be set.
     * @type {?}
     */
    GaEnhancedEcomImpressionFieldObject.prototype.name;
    /**
     * The list or collection to which the product belongs (e.g. Search Results)
     * @type {?}
     */
    GaEnhancedEcomImpressionFieldObject.prototype.list;
    /**
     * The brand associated with the product (e.g. Google).
     * @type {?}
     */
    GaEnhancedEcomImpressionFieldObject.prototype.brand;
    /**
     * The category to which the product belongs (e.g. Apparel).
     * Use / as a delimiter to specify up to 5-levels of hierarchy (e.g. Apparel/Men/T-Shirts)
     * @type {?}
     */
    GaEnhancedEcomImpressionFieldObject.prototype.category;
    /**
     * The variant of the product (e.g. Black).
     * @type {?}
     */
    GaEnhancedEcomImpressionFieldObject.prototype.variant;
    /**
     * The product's position in a list or collection (e.g. 2)
     * @type {?}
     */
    GaEnhancedEcomImpressionFieldObject.prototype.position;
    /**
     * The price of a product (e.g. 29.20)
     * @type {?}
     */
    GaEnhancedEcomImpressionFieldObject.prototype.price;
}
/**
 * Product Data
 *
 * Product data represents individual products that were viewed, added
 * to the shopping cart, etc. It is referred to as a productFieldObject
 * and contains the following values.
 *
 * @link https://developers.google.com/analytics/devguides/collection/analyticsjs/enhanced-ecommerce#product-data
 * @record
 */
export function GaEnhancedEcomProductFieldObject() { }
function GaEnhancedEcomProductFieldObject_tsickle_Closure_declarations() {
    /**
     * The product ID or SKU (e.g. P67890). *Either this field or name must be set.
     * @type {?}
     */
    GaEnhancedEcomProductFieldObject.prototype.id;
    /**
     * The name of the product (e.g. Android T-Shirt). *Either this field or id must be set.
     * @type {?}
     */
    GaEnhancedEcomProductFieldObject.prototype.name;
    /**
     * The brand associated with the product (e.g. Google).
     * @type {?}
     */
    GaEnhancedEcomProductFieldObject.prototype.brand;
    /**
     * The category to which the product belongs (e.g. Apparel).
     * Use / as a delimiter to specify up to 5-levels of hierarchy (e.g. Apparel/Men/T-Shirts).
     * @type {?}
     */
    GaEnhancedEcomProductFieldObject.prototype.category;
    /**
     * The variant of the product (e.g. Black).
     * @type {?}
     */
    GaEnhancedEcomProductFieldObject.prototype.variant;
    /**
     * The price of a product (e.g. 29.20).
     * @type {?}
     */
    GaEnhancedEcomProductFieldObject.prototype.price;
    /**
     * The quantity of a product (e.g. 2).
     * @type {?}
     */
    GaEnhancedEcomProductFieldObject.prototype.quantity;
    /**
     * The coupon code associated with a product (e.g. SUMMER_SALE13).
     * @type {?}
     */
    GaEnhancedEcomProductFieldObject.prototype.coupon;
    /**
     * The product's position in a list or collection (e.g. 2).
     * @type {?}
     */
    GaEnhancedEcomProductFieldObject.prototype.position;
}
/**
 * Promotion Data
 *
 * Represents information about a promotion that has been viewed.
 * It is referred to a promoFieldObject and contains the following values.
 *
 * @link https://developers.google.com/analytics/devguides/collection/analyticsjs/enhanced-ecommerce#promotion-data
 * @record
 */
export function GaEnhancedEcomPromoFieldObject() { }
function GaEnhancedEcomPromoFieldObject_tsickle_Closure_declarations() {
    /**
     * The promotion ID (e.g. PROMO_1234). *Either this field or name must be set.
     * @type {?}
     */
    GaEnhancedEcomPromoFieldObject.prototype.id;
    /**
     * The name of the promotion (e.g. Summer Sale). *Either this field or id must be set.
     * @type {?}
     */
    GaEnhancedEcomPromoFieldObject.prototype.name;
    /**
     * The creative associated with the promotion (e.g. summer_banner2).
     * @type {?}
     */
    GaEnhancedEcomPromoFieldObject.prototype.creative;
    /**
     * The position of the creative (e.g. banner_slot_1).
     * @type {?}
     */
    GaEnhancedEcomPromoFieldObject.prototype.position;
}
/**
 * Action Data
 *
 * Represents information about an ecommerce related action
 * that has taken place. It is referred to as an actionFieldObject
 * and contains the following values.
 *
 * @link https://developers.google.com/analytics/devguides/collection/analyticsjs/enhanced-ecommerce#action-data
 * @record
 */
export function GaEnhancedEcomActionFieldObject() { }
function GaEnhancedEcomActionFieldObject_tsickle_Closure_declarations() {
    /**
     * The transaction ID (e.g. T1234). *Required if the action type is purchase or refund.
     * @type {?}
     */
    GaEnhancedEcomActionFieldObject.prototype.id;
    /**
     * The store or affiliation from which this transaction occurred (e.g. Google Store).
     * @type {?}
     */
    GaEnhancedEcomActionFieldObject.prototype.affilation;
    /**
     * Specifies the total revenue or grand total associated with the transaction (e.g. 11.99).
     * This value may include shipping, tax costs, or other adjustments to total revenue that
     * you want to include as part of your revenue calculations. Note: if revenue is not set,
     * its value will be automatically calculated using the product quantity and price fields
     * of all products in the same hit.
     * @type {?}
     */
    GaEnhancedEcomActionFieldObject.prototype.revenue;
    /**
     * The total tax associated with the transaction.
     * @type {?}
     */
    GaEnhancedEcomActionFieldObject.prototype.tax;
    /**
     * The shipping cost associated with the transaction.
     * @type {?}
     */
    GaEnhancedEcomActionFieldObject.prototype.shipping;
    /**
     * The transaction coupon redeemed with the transaction.
     * @type {?}
     */
    GaEnhancedEcomActionFieldObject.prototype.coupon;
    /**
     * The list that the associated products belong to. Optional.
     * @type {?}
     */
    GaEnhancedEcomActionFieldObject.prototype.list;
    /**
     * A number representing a step in the checkout process. Optional on checkout actions.
     * @type {?}
     */
    GaEnhancedEcomActionFieldObject.prototype.step;
    /**
     * Additional field for checkout and checkout_option actions that can describe
     * option information on the checkout page, like selected payment method.
     * @type {?}
     */
    GaEnhancedEcomActionFieldObject.prototype.option;
}
