// angulartics2/ga/bundles/angulartics2-ga.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('angulartics2')) :
    typeof define === 'function' && define.amd ? define('angulartics2/ga', ['exports', '@angular/core', 'angulartics2'], factory) :
    (factory((global.angulartics2 = global.angulartics2 || {}, global.angulartics2.ga = {}),global.ng.core,global.angulartics2));
}(this, (function (exports,core,angulartics2) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    var __assign = function () {
        __assign = Object.assign || function __assign(t) {
            for (var s, i = 1, n = arguments.length; i < n; i++) {
                s = arguments[i];
                for (var p in s)
                    if (Object.prototype.hasOwnProperty.call(s, p))
                        t[p] = s[p];
            }
            return t;
        };
        return __assign.apply(this, arguments);
    };
    function __values(o) {
        var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
        if (m)
            return m.call(o);
        return {
            next: function () {
                if (o && i >= o.length)
                    o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var GoogleAnalyticsDefaults = (function () {
        function GoogleAnalyticsDefaults() {
            this.additionalAccountNames = [];
            this.userId = null;
            this.transport = '';
            this.anonymizeIp = false;
        }
        return GoogleAnalyticsDefaults;
    }());
    var Angulartics2GoogleAnalytics = (function () {
        function Angulartics2GoogleAnalytics(angulartics2$$1) {
            var _this = this;
            this.angulartics2 = angulartics2$$1;
            this.dimensionsAndMetrics = [];
            var /** @type {?} */ defaults = new GoogleAnalyticsDefaults();
            // Set the default settings for this module
            this.angulartics2.settings.ga = __assign({}, defaults, this.angulartics2.settings.ga);
            this.angulartics2.pageTrack
                .pipe(this.angulartics2.filterDeveloperMode())
                .subscribe(function (x) { return _this.pageTrack(x.path); });
            this.angulartics2.eventTrack
                .pipe(this.angulartics2.filterDeveloperMode())
                .subscribe(function (x) { return _this.eventTrack(x.action, x.properties); });
            this.angulartics2.exceptionTrack
                .pipe(this.angulartics2.filterDeveloperMode())
                .subscribe(function (x) { return _this.exceptionTrack(x); });
            this.angulartics2.setUsername.subscribe(function (x) { return _this.setUsername(x); });
            this.angulartics2.setUserProperties.subscribe(function (x) { return _this.setUserProperties(x); });
            this.angulartics2.userTimings
                .pipe(this.angulartics2.filterDeveloperMode())
                .subscribe(function (x) { return _this.userTimings(x); });
        }
        /**
         * @param {?} path
         * @return {?}
         */
        Angulartics2GoogleAnalytics.prototype.pageTrack = /**
         * @param {?} path
         * @return {?}
         */
            function (path) {
                if (typeof _gaq !== 'undefined' && _gaq) {
                    _gaq.push(['_trackPageview', path]);
                    try {
                        for (var _a = __values(this.angulartics2.settings.ga.additionalAccountNames), _b = _a.next(); !_b.done; _b = _a.next()) {
                            var accountName = _b.value;
                            _gaq.push([accountName + '._trackPageview', path]);
                        }
                    }
                    catch (e_1_1) {
                        e_1 = { error: e_1_1 };
                    }
                    finally {
                        try {
                            if (_b && !_b.done && (_c = _a.return))
                                _c.call(_a);
                        }
                        finally {
                            if (e_1)
                                throw e_1.error;
                        }
                    }
                }
                if (typeof ga !== 'undefined' && ga) {
                    if (this.angulartics2.settings.ga.userId) {
                        ga('set', '&uid', this.angulartics2.settings.ga.userId);
                        try {
                            for (var _d = __values(this.angulartics2.settings.ga.additionalAccountNames), _e = _d.next(); !_e.done; _e = _d.next()) {
                                var accountName = _e.value;
                                ga(accountName + '.set', '&uid', this.angulartics2.settings.ga.userId);
                            }
                        }
                        catch (e_2_1) {
                            e_2 = { error: e_2_1 };
                        }
                        finally {
                            try {
                                if (_e && !_e.done && (_f = _d.return))
                                    _f.call(_d);
                            }
                            finally {
                                if (e_2)
                                    throw e_2.error;
                            }
                        }
                    }
                    if (this.angulartics2.settings.ga.anonymizeIp) {
                        ga('set', 'anonymizeIp', true);
                        try {
                            for (var _g = __values(this.angulartics2.settings.ga.additionalAccountNames), _h = _g.next(); !_h.done; _h = _g.next()) {
                                var accountName = _h.value;
                                ga(accountName + '.set', 'anonymizeIp', true);
                            }
                        }
                        catch (e_3_1) {
                            e_3 = { error: e_3_1 };
                        }
                        finally {
                            try {
                                if (_h && !_h.done && (_j = _g.return))
                                    _j.call(_g);
                            }
                            finally {
                                if (e_3)
                                    throw e_3.error;
                            }
                        }
                    }
                    ga('send', 'pageview', path);
                    try {
                        for (var _k = __values(this.angulartics2.settings.ga.additionalAccountNames), _l = _k.next(); !_l.done; _l = _k.next()) {
                            var accountName = _l.value;
                            ga(accountName + '.send', 'pageview', path);
                        }
                    }
                    catch (e_4_1) {
                        e_4 = { error: e_4_1 };
                    }
                    finally {
                        try {
                            if (_l && !_l.done && (_m = _k.return))
                                _m.call(_k);
                        }
                        finally {
                            if (e_4)
                                throw e_4.error;
                        }
                    }
                }
                var e_1, _c, e_2, _f, e_3, _j, e_4, _m;
            };
        /**
         * Track Event in GA
         *
         * @param action Associated with the event
         * @param properties Comprised of:
         *  - category (string) and optional
         *  - label (string)
         *  - value (integer)
         *  - noninteraction (boolean)
         *
         * @link https://developers.google.com/analytics/devguides/collection/gajs/eventTrackerGuide#SettingUpEventTracking
         * @link https://developers.google.com/analytics/devguides/collection/analyticsjs/events
         */
        /**
         * Track Event in GA
         *
         * @link https://developers.google.com/analytics/devguides/collection/gajs/eventTrackerGuide#SettingUpEventTracking
         * @link https://developers.google.com/analytics/devguides/collection/analyticsjs/events
         * @param {?} action Associated with the event
         * @param {?} properties Comprised of:
         *  - category (string) and optional
         *  - label (string)
         *  - value (integer)
         *  - noninteraction (boolean)
         *
         * @return {?}
         */
        Angulartics2GoogleAnalytics.prototype.eventTrack = /**
         * Track Event in GA
         *
         * @link https://developers.google.com/analytics/devguides/collection/gajs/eventTrackerGuide#SettingUpEventTracking
         * @link https://developers.google.com/analytics/devguides/collection/analyticsjs/events
         * @param {?} action Associated with the event
         * @param {?} properties Comprised of:
         *  - category (string) and optional
         *  - label (string)
         *  - value (integer)
         *  - noninteraction (boolean)
         *
         * @return {?}
         */
            function (action, properties) {
                // Google Analytics requires an Event Category
                if (!properties || !properties.category) {
                    properties = properties || {};
                    properties.category = 'Event';
                }
                // GA requires that eventValue be an integer, see:
                // https://developers.google.com/analytics/devguides/collection/analyticsjs/field-reference#eventValue
                // https://github.com/luisfarzati/angulartics/issues/81
                if (properties.value) {
                    var /** @type {?} */ parsed = parseInt(properties.value, 10);
                    properties.value = isNaN(parsed) ? 0 : parsed;
                }
                if (typeof ga !== 'undefined') {
                    var /** @type {?} */ eventOptions = {
                        eventCategory: properties.category,
                        eventAction: action,
                        eventLabel: properties.label,
                        eventValue: properties.value,
                        nonInteraction: properties.noninteraction,
                        page: properties.page || location.hash.substring(1) || location.pathname,
                        userId: this.angulartics2.settings.ga.userId,
                        hitCallback: properties.hitCallback,
                    };
                    // add custom dimensions and metrics
                    this.setDimensionsAndMetrics(properties);
                    if (this.angulartics2.settings.ga.transport) {
                        ga('send', 'event', eventOptions, {
                            transport: this.angulartics2.settings.ga.transport,
                        });
                    }
                    else {
                        ga('send', 'event', eventOptions);
                    }
                    try {
                        for (var _a = __values(this.angulartics2.settings.ga.additionalAccountNames), _b = _a.next(); !_b.done; _b = _a.next()) {
                            var accountName = _b.value;
                            ga(accountName + '.send', 'event', eventOptions);
                        }
                    }
                    catch (e_5_1) {
                        e_5 = { error: e_5_1 };
                    }
                    finally {
                        try {
                            if (_b && !_b.done && (_c = _a.return))
                                _c.call(_a);
                        }
                        finally {
                            if (e_5)
                                throw e_5.error;
                        }
                    }
                }
                else if (typeof _gaq !== 'undefined') {
                    _gaq.push([
                        '_trackEvent',
                        properties.category,
                        action,
                        properties.label,
                        properties.value,
                        properties.noninteraction,
                    ]);
                }
                var e_5, _c;
            };
        /**
         * Exception Track Event in GA
         *
         * @param properties Comprised of the optional fields:
         *  - fatal (string)
         *  - description (string)
         *
         * @https://developers.google.com/analytics/devguides/collection/analyticsjs/exceptions
         *
         * @link https://developers.google.com/analytics/devguides/collection/analyticsjs/events
         */
        /**
         * Exception Track Event in GA
         *
         * \@https://developers.google.com/analytics/devguides/collection/analyticsjs/exceptions
         *
         * @link https://developers.google.com/analytics/devguides/collection/analyticsjs/events
         * @param {?} properties Comprised of the optional fields:
         *  - fatal (string)
         *  - description (string)
         *
         * @return {?}
         */
        Angulartics2GoogleAnalytics.prototype.exceptionTrack = /**
         * Exception Track Event in GA
         *
         * \@https://developers.google.com/analytics/devguides/collection/analyticsjs/exceptions
         *
         * @link https://developers.google.com/analytics/devguides/collection/analyticsjs/events
         * @param {?} properties Comprised of the optional fields:
         *  - fatal (string)
         *  - description (string)
         *
         * @return {?}
         */
            function (properties) {
                if (properties.fatal === undefined) {
                    console.log('No "fatal" provided, sending with fatal=true');
                    properties.fatal = true;
                }
                properties.exDescription = properties.description;
                var /** @type {?} */ eventOptions = {
                    exFatal: properties.fatal,
                    exDescription: properties.description,
                };
                ga('send', 'exception', eventOptions);
                try {
                    for (var _a = __values(this.angulartics2.settings.ga.additionalAccountNames), _b = _a.next(); !_b.done; _b = _a.next()) {
                        var accountName = _b.value;
                        ga(accountName + '.send', 'exception', eventOptions);
                    }
                }
                catch (e_6_1) {
                    e_6 = { error: e_6_1 };
                }
                finally {
                    try {
                        if (_b && !_b.done && (_c = _a.return))
                            _c.call(_a);
                    }
                    finally {
                        if (e_6)
                            throw e_6.error;
                    }
                }
                var e_6, _c;
            };
        /**
         * User Timings Event in GA
         * @name userTimings
         *
         * @param properties Comprised of the mandatory fields:
         *  - timingCategory (string)
         *  - timingVar (string)
         *  - timingValue (number)
         * Properties can also have the optional fields:
         *  - timingLabel (string)
         *
         * @link https://developers.google.com/analytics/devguides/collection/analyticsjs/user-timings
         */
        /**
         * User Timings Event in GA
         * \@name userTimings
         *
         * @link https://developers.google.com/analytics/devguides/collection/analyticsjs/user-timings
         * @param {?} properties Comprised of the mandatory fields:
         *  - timingCategory (string)
         *  - timingVar (string)
         *  - timingValue (number)
         * Properties can also have the optional fields:
         *  - timingLabel (string)
         *
         * @return {?}
         */
        Angulartics2GoogleAnalytics.prototype.userTimings = /**
         * User Timings Event in GA
         * \@name userTimings
         *
         * @link https://developers.google.com/analytics/devguides/collection/analyticsjs/user-timings
         * @param {?} properties Comprised of the mandatory fields:
         *  - timingCategory (string)
         *  - timingVar (string)
         *  - timingValue (number)
         * Properties can also have the optional fields:
         *  - timingLabel (string)
         *
         * @return {?}
         */
            function (properties) {
                if (!properties ||
                    !properties.timingCategory ||
                    !properties.timingVar ||
                    !properties.timingValue) {
                    console.error('Properties timingCategory, timingVar, and timingValue are required to be set.');
                    return;
                }
                if (typeof ga !== 'undefined') {
                    ga('send', 'timing', properties);
                    try {
                        for (var _a = __values(this.angulartics2.settings.ga.additionalAccountNames), _b = _a.next(); !_b.done; _b = _a.next()) {
                            var accountName = _b.value;
                            ga(accountName + '.send', 'timing', properties);
                        }
                    }
                    catch (e_7_1) {
                        e_7 = { error: e_7_1 };
                    }
                    finally {
                        try {
                            if (_b && !_b.done && (_c = _a.return))
                                _c.call(_a);
                        }
                        finally {
                            if (e_7)
                                throw e_7.error;
                        }
                    }
                }
                var e_7, _c;
            };
        /**
         * @param {?} userId
         * @return {?}
         */
        Angulartics2GoogleAnalytics.prototype.setUsername = /**
         * @param {?} userId
         * @return {?}
         */
            function (userId) {
                this.angulartics2.settings.ga.userId = userId;
                if (typeof ga === 'undefined') {
                    return;
                }
                ga('set', 'userId', userId);
            };
        /**
         * @param {?} properties
         * @return {?}
         */
        Angulartics2GoogleAnalytics.prototype.setUserProperties = /**
         * @param {?} properties
         * @return {?}
         */
            function (properties) {
                this.setDimensionsAndMetrics(properties);
            };
        /**
         * @param {?} properties
         * @return {?}
         */
        Angulartics2GoogleAnalytics.prototype.setDimensionsAndMetrics = /**
         * @param {?} properties
         * @return {?}
         */
            function (properties) {
                var _this = this;
                if (typeof ga === 'undefined') {
                    return;
                }
                // clean previously used dimensions and metrics that will not be overriden
                this.dimensionsAndMetrics.forEach(function (elem) {
                    if (!properties.hasOwnProperty(elem)) {
                        ga('set', elem, undefined);
                        _this.angulartics2.settings.ga.additionalAccountNames.forEach(function (accountName) {
                            ga(accountName + ".set", elem, undefined);
                        });
                    }
                });
                this.dimensionsAndMetrics = [];
                // add custom dimensions and metrics
                Object.keys(properties).forEach(function (key) {
                    if (key.lastIndexOf('dimension', 0) === 0 ||
                        key.lastIndexOf('metric', 0) === 0) {
                        ga('set', key, properties[key]);
                        _this.angulartics2.settings.ga.additionalAccountNames.forEach(function (accountName) {
                            ga(accountName + ".set", key, properties[key]);
                        });
                        _this.dimensionsAndMetrics.push(key);
                    }
                });
            };
        Angulartics2GoogleAnalytics.decorators = [
            { type: core.Injectable }
        ];
        /** @nocollapse */
        Angulartics2GoogleAnalytics.ctorParameters = function () {
            return [
                { type: angulartics2.Angulartics2 }
            ];
        };
        return Angulartics2GoogleAnalytics;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    exports.GoogleAnalyticsDefaults = GoogleAnalyticsDefaults;
    exports.Angulartics2GoogleAnalytics = Angulartics2GoogleAnalytics;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
