// angulartics2/uiroutermodule/bundles/angulartics2-uiroutermodule.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@uirouter/core'), require('rxjs'), require('angulartics2')) :
    typeof define === 'function' && define.amd ? define('angulartics2/uiroutermodule', ['exports', '@angular/core', '@uirouter/core', 'rxjs', 'angulartics2'], factory) :
    (factory((global.angulartics2 = global.angulartics2 || {}, global.angulartics2.uiroutermodule = {}),global.ng.core,null,global.rxjs,global.angulartics2));
}(this, (function (exports,core,core$1,rxjs,angulartics2) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    /**
     * Track Route changes for applications using UI-Router
     *
     * @link https://ui-router.github.io/ng2/docs/latest/
     *
     * referenced: https://github.com/ui-router/sample-app-angular/blob/9adb533b85c0f0fccef23968489cca0a5ec84654/src/app/util/ga.ts
     */
    var UIRouterTracking = (function () {
        function UIRouterTracking(transitionService) {
            this.transitionService = transitionService;
        }
        /**
         * @param {?} trans
         * @return {?}
         */
        UIRouterTracking.prototype.path = /**
         * @param {?} trans
         * @return {?}
         */
            function (trans) {
                return trans.$to().url.format(trans.params());
            };
        /**
         * @param {?} settings
         * @return {?}
         */
        UIRouterTracking.prototype.trackLocation = /**
         * @param {?} settings
         * @return {?}
         */
            function (settings) {
                var _this = this;
                var /** @type {?} */ subject = new rxjs.Subject();
                this.transitionService.onSuccess({}, function (trans) {
                    return subject.next({ url: _this.path(trans) });
                }, {
                    priority: -10000,
                });
                return subject;
            };
        /**
         * @param {?} url
         * @return {?}
         */
        UIRouterTracking.prototype.prepareExternalUrl = /**
         * @param {?} url
         * @return {?}
         */
            function (url) {
                return url;
            };
        UIRouterTracking.decorators = [
            { type: core.Injectable }
        ];
        /** @nocollapse */
        UIRouterTracking.ctorParameters = function () {
            return [
                { type: core$1.TransitionService }
            ];
        };
        return UIRouterTracking;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var Angulartics2UirouterModule = (function () {
        function Angulartics2UirouterModule() {
        }
        /**
         * @param {?} providers
         * @param {?=} settings
         * @return {?}
         */
        Angulartics2UirouterModule.forRoot = /**
         * @param {?} providers
         * @param {?=} settings
         * @return {?}
         */
            function (providers, settings) {
                if (settings === void 0) {
                    settings = {};
                }
                return {
                    ngModule: Angulartics2UirouterModule,
                    providers: __spread([
                        { provide: angulartics2.ANGULARTICS2_TOKEN, useValue: { providers: providers, settings: settings } },
                        angulartics2.Angulartics2,
                        { provide: angulartics2.RouterlessTracking, useClass: UIRouterTracking }
                    ], providers),
                };
            };
        Angulartics2UirouterModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [angulartics2.Angulartics2OnModule],
                    },] }
        ];
        return Angulartics2UirouterModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    exports.Angulartics2UirouterModule = Angulartics2UirouterModule;
    exports.UIRouterTracking = UIRouterTracking;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
